src/pauli_exp_val.rs

```diff
@@ -0,0 +1,193 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::convert::TryInto;
+
+use num_complex::Complex64;
+use numpy::PyReadonlyArray1;
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use rayon::prelude::*;
+
+use crate::eval_parallel_env;
+
+const LANES: usize = 8;
+
+#[inline]
+fn fast_sum(values: &[f64]) -> f64 {
+    let chunks = values.chunks_exact(LANES);
+    let remainder = chunks.remainder();
+
+    let sum = chunks.fold([0.; LANES], |mut acc, chunk| {
+        let chunk: [f64; LANES] = chunk.try_into().unwrap();
+        for i in 0..LANES {
+            acc[i] += chunk[i];
+        }
```

Comment:
Yeah exactly, the default `iter().sum()` was noticeably slower than the cython version with generic CPU tuning. When I set the target to either `native` or `x86_64-v(2|3)`the sum was significantly faster because it was using AVX or AVX2 but that isn't viable if we want to support pre-2008 CPUs. So I went searching for how to efficiently do SIMD based addition in Rust and chunking it to a fixed size like this enables the compiler to use SSE instructions since it's <= 512bits and fixed width. I actually based this off of https://stackoverflow.com/a/67191480/14033130 which provided the example for f32. Which reminds me I forgot to provide attribution to the source, I added that in: https://github.com/Qiskit/qiskit-terra/pull/7702/commits/250d508f73bea329d034623d50de1b8481042cb6